x/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,198 @@
+			Height:        4,
+			ChainId:       testApp.GetChainID(),
+			BlockData:     &tmproto.Data{},
+			BlockDataSize: 1e6,
+		})
+
+		require.Len(t, resp.BlockData.Txs, 1)
+		tx, err := testApp.GetTxConfig().TxDecoder()(resp.BlockData.Txs[0])
+		require.NoError(t, err)
+		require.Len(t, tx.GetMsgs(), 1)
+		msg, ok := tx.GetMsgs()[0].(*upgrade.MsgVersionChange)
+		require.True(t, ok)
+		require.EqualValues(t, 2, msg.Version)
+	}
+
+	{
+		// Height 5 however is outside the range and thus the upgrade
+		// message should not be prepended
+		resp := testApp.PrepareProposal(abci.RequestPrepareProposal{
+			Height:        5,
+			ChainId:       testApp.GetChainID(),
+			BlockData:     &tmproto.Data{},
+			BlockDataSize: 1e6,
+		})
+		require.Len(t, resp.BlockData.Txs, 0)
+	}
+
+	// We should accept this proposal as valid
+	processProposalResp := testApp.ProcessProposal(abci.RequestProcessProposal{
+		Header: tmproto.Header{
+			Height:   2,
+			DataHash: resp.BlockData.Hash,
+		},
+		BlockData: resp.BlockData,
+	})
+	require.True(t, processProposalResp.IsOK())
+
+	{
+		// to assert that the upgrade tx must be the first tx
+		// we insert a tx before the upgrade tx. To get the hash
+		// we need to build the square and data availability header
+		txs := [][]byte{[]byte("hello world"), upgradeTx}
+		dataSquare, txs, err := square.Build(txs, appconsts.LatestVersion, appconsts.DefaultGovMaxSquareSize)
+		require.NoError(t, err)
+		eds, err := da.ExtendShares(shares.ToBytes(dataSquare))
+		require.NoError(t, err)
+		dah, err := da.NewDataAvailabilityHeader(eds)
+		require.NoError(t, err)
+		blockData := &tmproto.Data{
+			Txs:        txs,
+			SquareSize: uint64(dataSquare.Size()),
+			Hash:       dah.Hash(),
+		}
+
+		processProposalResp := testApp.ProcessProposal(abci.RequestProcessProposal{
+			Header: tmproto.Header{
+				Height:   2,
+				DataHash: blockData.Hash,
+			},
+			BlockData: blockData,
+		})
+		require.True(t, processProposalResp.IsRejected())
+	}
+
+	testApp.BeginBlock(abci.RequestBeginBlock{Header: tmproto.Header{Height: 2}})
+	respDeliverTx := testApp.DeliverTx(abci.RequestDeliverTx{Tx: resp.BlockData.Txs[0]})
+	require.EqualValues(t, 0, respDeliverTx.Code, respDeliverTx.Log)
+	// app version should not have changed yet
+	require.EqualValues(t, 1, testApp.AppVersion())
+	respEndBlock := testApp.EndBlock(abci.RequestEndBlock{Height: 2})
+	// now the app version changes
+	require.EqualValues(t, 2, respEndBlock.ConsensusParamUpdates.Version.AppVersion)
+	require.EqualValues(t, 2, testApp.AppVersion())
+
+	_ = testApp.Commit()
+
+	// If another node proposes a block with a version change that is
+	// not supported by the nodes own state machine then the node
+	// rejects the proposed block
+	respProcessProposal := testApp.ProcessProposal(abci.RequestProcessProposal{
+		Header: tmproto.Header{
+			Height: 3,
+		},
+		BlockData: &tmproto.Data{
+			Txs: [][]byte{upgradeTx},
+		},
+	})
+	require.True(t, respProcessProposal.IsRejected())
```

Comment:
[question] why is this `upgradeTx` proposing a version change that is not supported by the nodes own state machine? It looks like this `upgradeTx` is proposing an upgrade to version 3. 

Is it because the test app is setup with the schedule: `upgrade.NewSchedule(upgrade.NewPlan(3, 5, 2))` which means the app only supports version 1 and 2?

Reply:
This is ProcessProposal so it's mimicking receiving a proposal from another node (presumably with v3) capabilities. This is just to check that it will reject the proposal